### ISSUE TITLE
A "sprite" object always faces north, and so do its attachments.

### ISSUE THREAD

**Issue by hecktest** [minetest/minetest#9191]
5.1.0
If an object has a "sprite" visual, it always faces north and ignores any attempts to set rotation.
An unpleasant side effect of this is that objects attached to it will always face north as well.
![oops](https://user-images.githubusercontent.com/42101236/70740291-2166e200-1d19-11ea-9099-5ca1a3c31494.gif)

**Comment by rubenwardy** [comment]
Hi, please place any nsfw imagery in details tags for the time being, like so:

```html
<details>
<summary>NSFW</summary>
Image
</details>
```

Our policy to this may change in the future, I have raised a concern with other staff members. Such imagery is unprofessional, risky for those in public, and generally unwelcoming to any female contributors

**Comment by paramat** [comment]
Does not seem NSFW, no nudity.
If someone posted images of a super-masculine male player model in nothing but a loincloth (many fantasy games have something like this), men would not feel unwelcome, so the 'offence to women' claim is a mistake. Please let us wait for any women to express any offence.

**Comment by rubenwardy** [comment]
Mate, a scantly clad anime girl is not work or public safe whatsoever. I'm not sure what games you're playing that contain women in thongs, but I've certainly haven't come across any

**Comment by Moth-Tolias** [comment]
not offended but would appreciate the warning, yeah.
_-_ a woman

**Comment by Warr1024** [comment]
Most of the issues featuring these kinds of images could also have been demonstrated using MTG models and minimal example mods, which would also make them easier for devs to test, reproduce, and verify any potential fix.

For example, when I submitted #9029, I didn't use the more complex textures I created that caused me to discover the issue, but instead made a simplified example mod that clearly shows the specific issues and minimizes factors in the bug report that are not relevant to the problem.  Going through this exercise also confirms that it really _is_ a reproducible engine bug and not a problem specific to anything I've created.

This is probably a good policy to follow _regardless_ of the specific content involved, and would probably minimize the risk of unrelated issues impeding the bugfix.

**Comment by hecktest** [comment]
If you want me to jump through extra hoops (remaking the whole test case in minimal) in order to submit bugs, because you're offended by something that _isn't considered obscene in most of the civilized world_, then I will stop reporting bugs. In the case of making repros, you are asking me to spend half a day on something that used to take ten minutes. I require a nude texture in order to have customizable clothing; and I simply have not begun work on the clothing textures. This is to avoid wasting work when some issues with the model surface; _this has happened before!_ Some of you fine folks using the IRC might even remember the old model.

This is also concern trolling at its finest, you're basically being offended in advance, for the sake of some imaginary visitor who might complain about this, even though they haven't yet. No one has until now, and it's been a while.
The concern is also inflated; underwear is obscene now? Do you also avoid beaches? If you haven't come across a game that contains this, then man, you've got some catching up to do. Take a stroll through any digital game store, paying extra attention to all the open world and eastern productions specifically. After you do that, go and complain to all the developers who made those, and all the people on the internet doing free marketing for them. Then realize you're trying to bully someone without even a fraction of the resources they have. Who's just reporting bugs. While in the official game you can bludgeon a fluffy rabbit to death with a pick axe, splattering its blood everywhere, then skin, cook and eat it, and every other server hosts dozens of texture skins featuring people in their underwear, cute girls included.

Don't you have better things to do than getting your panties in a bunch over, well, panties? Goodness.

**Comment by LoneWolfHT** [comment]
I'm fine if you put it in a spoiler. I'd rather not have those kinds of images suddenly pop up while I'm scrolling my email. I have young kids looking over my shoulder sometimes.
If you felt like putting a bit more time into your bug reports and started reproducing with minimal I would 100% support that too

- an 'offended' person (7/8 sarcasm there)

**Comment by SmallJoker** [comment]
Does it work after https://github.com/minetest/minetest/commit/aa8df11#diff-90993ca3bb550ffc1c77a1578f3545d3R591 respectively 103d9c5 ?

**Comment by rubenwardy** [comment]
There is no imaginary visitor, **I** am already affected by this in the fact that I can't open any issues by you in public. I'm literally just asking you to use spoiler tags so people can choose whether it's an appropriate time to view the images. It's in your benefit, as you will receive responses faster

**Comment by jastevenson303** [comment]
Lol.

**Comment by hecktest** [comment]
> 
> 
> Does it work after [aa8df11#diff-90993ca3bb550ffc1c77a1578f3545d3R591](https://github.com/minetest/minetest/commit/aa8df11#diff-90993ca3bb550ffc1c77a1578f3545d3R591) respectively [103d9c5](https://github.com/minetest/minetest/commit/103d9c5c536abd15a79eb7ed899fee29168bb14c) ?

Nope, sorry. But it seems like another client environment issue; get_yaw returns the previously set angle as expected.
![hiwardy](https://user-images.githubusercontent.com/42101236/70965505-1fda4880-2090-11ea-825b-e234512a6bbc.gif)

**Comment by Wuzzy2** [comment]
I don't really think if I understand the problem here. Your GIFs don't seem to indicate what part of it is the actual sprite.

According to `lua_api.txt`: 

> "sprite" is a flat texture always facing the player.

But nothing in your GIFs seem to indicate that. I am confused.

Can you please post the full entity definition that you have been showing in your GIFs?

**Comment by Wuzzy2** [comment]
@rubenwardy: Perhaps hecktest doesn't even know how the “spoiler tag” works? I cannot find it even in <https://github.github.com/gfm/>. :-(

**Comment by Wuzzy2** [comment]
No matter what I do, I just cannot reproduce the supposed bug.
As I said previously, it's intentional the `sprite` visual is always facing the player. And you're right, no rotation will ever affect its appearance. I believe this is by design.

Maybe you want to use `upright_sprite` instead. In my testing, this sprite can be rotated with `set_rotation` along all 3 axes. Note that `upright_sprite` uses 2 textures, not 1.

**Comment by Wuzzy2** [comment]
@hecktest: I'm afraid nobody can help with fixing this issue without giving more details.

Your issue description is too vague to be useful. Please give us:

* The exact version number
* The exact steps to reproduce
* The full entity definitions (both “parent” and “child”)

There is now a full month since your last reply, and it's still not reproducible.

**Comment by SmallJoker** [comment]
Or even better: a simple testing mod to reproduce this issue. I'm not going two write a testing mod for each bug, thus it be very welcome in order to figure out what's going wrong here.

**Comment by Wuzzy2** [comment]
**I REPRODUCED THE BUG!!!**
I guess …

## Proper bug description

If you attach an object to an object with the `sprite` visual, then try to rotate the sprite object, the rotation is ignored for both the sprite object and the attached object.
However, the expected behavior is apparently that the attached object “inherits” the rotation of its parent. If the parent object is any other visual (e.g. `cube`), the child will rotate as the parent rotates.

## Steps to reproduce (generic)

What you need to do:

1. Spawn parent entity: An entity with `visual="sprite"`
2. Spawn child entity: Just any entity will do, it just must be visible
3. Attach child entity to parent entity
4. Rotate parent entity

As the parent is using the sprite visual, it ignored all attempts of rotation. This is expected. HOWEVER the child entity will ignore the rotation as well.

For comparison, do the same steps, but use any other visual (e.g. "cube") for the parent. Now the rotation will both affect parent AND child. It seems the intended behavior is that the child “inherits” the rotation from the parent.

## Steps to reproduce (`devtest`)

To helping you reproduce this, I have today just added a couple more tools to `devtest` only for this bug.

How to reproduce it in `devtest`:

1. Spawn the entity `testentities:sprite` (=parent)
2. Go somewhere else and spawn `testentities:cube` (=child)
3. Use the object attacher. Punch the sprite entity, then the cube (if you are lost, look into `README.md` in the `testtools` mod)
4. Use the entity rotator and punch the sprite. 

For the comparison, repeat this test, but use `testentities:upright_sprite` instead of `testentities:sprite`. Now the rotation should be inherited properly by the cube.

Note: The object attacher calls `child:set_attach(parent, "", position, rotation)`, with `position` defaulting to `{x=0, y=11, z=0}` and `rotation` defaulting to `{x=0, y=0, z=0}`.

## Remaining confusion

I am still not sure if this is actually the same bug that OP reported. The GIF still does not match the description. Where in the image is the sprite? Is it the shield? But that doesn't make sense, because it's clearly not the parent object. Is the sprite the girl. Maybe …? Only the stars and OP know.

Also, OP claims that the `sprite` visual always faces North. I can not confirm this claim. For me, `sprite` visual always faces the player. I guess this was a miscommunication.

Frankly, this issue is very poorly worded and I had to read OP's mind to figure out what's going on. :-1:

**Comment by hecktest** [comment]
Visually it always faces the player, but transform wise it has no rotation, and always faces north with its +Z axis. The expected behavior is that despite the billboard visual, it would still handle attachments like any other object. Visual properties should not affect transforms.

**Comment by SmallJoker** [comment]
Until now I assumed you meant upright sprites, which do have a transformation matrix.
Sprites alone will indeed always face the player; it's how they work. Simple texture, no rotation. Use upright sprites for anything that needs to rotate.

**Comment by Wuzzy2** [comment]
Ummmm, did you read and understand my comment? I confirmed the bug, were you able to reproduce it?
